src/main/scala/mesosphere/marathon/core/externalvolume/impl/providers/DVDIProvider.scala

```diff
@@ -25,86 +25,67 @@ private[impl] case object DVDIProvider extends ExternalVolumeProvider {
-      val offset = i.filter(_.getName.startsWith(dvdiVolumeName)).map{ s =>
-        val ss = s.getName.substring(dvdiVolumeName.length)
-        if (ss.length > 0) ss.toInt else 0
-      }.foldLeft(-1)((z, i) => if (i > z) i else z)
-
-      if (offset >= 0) (offset + 1).toString else ""
-    }
-
-    def mkVar(name: String, value: String): Environment.Variable =
-      Environment.Variable.newBuilder.setName(name).setValue(value).build
-
-    val vars = Seq[Environment.Variable](
-      mkVar(dvdiVolumeContainerPath + suffix, vol.containerPath),
-      mkVar(dvdiVolumeName + suffix, vol.external.name),
-      mkVar(dvdiVolumeDriver + suffix, vol.external.options(driverOption))
-    )
-
-    val optsVar = {
+    def dockerVolumeParameters(volume: ExternalVolume) = {
+      import OptionLabelPatterns._
       val prefix: String = name + OptionNamespaceSeparator
       // don't let the user override these
       val ignore = Set(driverOption)
       // external.size trumps any user-specified dvdi/size option
-      val opts = vol.external.options ++ Map[String, String](
-        vol.external.size.map(prefix + "size" -> _.toString).toList: _*
+      val opts = volume.external.options ++ Map[String, String](
+        volume.external.size.map(prefix + "size" -> _.toString).toList: _*
       )
 
       // forward all dvdi/* options to the dvdcli driver, stripping the dvdi/ prefix
       // and trimming the values
       opts.filterKeys{ k =>
         k.startsWith(prefix) && !ignore.contains(k.toLowerCase)
       }.map{
-        case (k, v) => k.substring(prefix.length) + "=" + v.trim()
-      }.mkString(",")
+        case (k, v) => Parameter.newBuilder
+          .setKey(k.substring(prefix.length))
+          .setValue(v.trim())
+          .build
+      }.asJava
+    }
+
+    def toUnifiedMesosVolume(volume: ExternalVolume): MesosVolume = {
+      val driverName = volume.external.options(driverOption)
+      MesosVolume.newBuilder
+        .setContainerPath(volume.containerPath)
+        .setMode(volume.mode)
+        .setSource(MesosVolume.Source.newBuilder
+          .setType(MesosVolume.Source.Type.DOCKER_VOLUME)
+          .setDockerVolume(MesosVolume.Source.DockerVolume.newBuilder
+            .setDriver(driverName)
+            .setName(volume.external.name)
+            .setDriverOptions(Parameters.newBuilder.addAllParameter(dockerVolumeParameters(volume)).build)
+            .build)
+          .build)
+        .build
     }
 
-    if (optsVar.isEmpty) vars
-    else { vars :+ mkVar(dvdiVolumeOpts + suffix, optsVar) }
+    // special behavior for docker vs. mesos containers
+    // - docker containerizer: serialize volumes into mesos proto
+    // - docker containerizer: specify "volumeDriver" for the container
+    if (builder.getType == ContainerInfo.Type.DOCKER && builder.hasDocker) {
+      val driverName = ev.external.options(driverOption)
+      builder.setDocker(builder.getDocker.toBuilder.setVolumeDriver(driverName).build)
```

Comment:
add a comment here along the lines of: because we're in docker-containerizer mode we won't be using the docker-volume-isolator. so we build "unsourced" plain volumes in this case